windows/src/engine/keyman32/kmprocess.cpp

```diff
@@ -142,27 +173,20 @@ BOOL ProcessHook()
 	}
 
   if (isUsingCoreProcessor) {  // TODO: 5442 Note: Nested if will be reduced once using core only
-    if (!_td->TIPFUpdateable) {
-      PWSTR contextBuf = _td->app->ContextBufMax(MAXCONTEXT);
-      km_kbp_context_item *citems = nullptr;
-      ContextItemsFromAppContext(contextBuf, &citems);
-      if (KM_KBP_STATUS_OK !=
-        (km_kbp_status_codes)km_kbp_context_set(
-          km_kbp_state_context(_td->lpActiveKeyboard->lpCoreKeyboardState), citems)) {
-        km_kbp_context_items_dispose(citems);
-        return FALSE;
-      }
-      km_kbp_context_items_dispose(citems);
-      SendDebugMessageFormat(0, sdmGlobal, 0, "ProcessEvent: vkey[%d] ShiftState[%d] isDown[%d]", _td->state.vkey,
-                                                    static_cast<uint16_t>(Globals::get_ShiftState() & K_MODIFIERFLAG),
-                                                    (uint8_t)_td->state.isDown);
-      if (KM_KBP_STATUS_OK !=
-        (km_kbp_status_codes)km_kbp_process_event(_td->lpActiveKeyboard->lpCoreKeyboardState, _td->state.vkey,
-                                  static_cast<uint16_t>(Globals::get_ShiftState() & K_MODIFIERFLAG),
-                                                    (uint8_t)_td->state.isDown)) {
-        SendDebugMessageFormat(0, sdmGlobal, 0, "ProcessEvent CoreProcessEvent Result:False %d ",FALSE);
+    // For applications not using the TSF kmtip calls this function twice for each keystroke,
+    // first to determine if we are doing processing work (TIPFUpdateable == FALSE),
+    // if we say yes it will call a second time to actually do the work.
+    // We call the core process event only once and use the core's queued actions
+    // on the second pass.
+    // For the TSF in most cases kmtip (except OnPreservedKey) will not call the non-updateable test pass.
+    // Therfore the core process event will need to be called before processing the actions.
+    if (!_td->TIPFUpdateable || !_td->CoreProcessEventRun) {
```

Comment:
@marc I was lying in bed the other night and thought hey I don't really even need to check if `!_td->TIPFUpdateable` in this if statement because the coreProcessEventRun is not set then that is sufficient reason to run.
I know it is sad that I was lying in bed thinking about it.  Anyway part of me wants to leave it there just to make sure. It also helps the developer follow what is going on but I did leave a pretty massive comment. Your thoughts?

Reply:
I like the explicitness of the test. Perhaps we should add a comment relating to your in-bed thoughts just for future code readers?